### PR TITLE
Migrate workbox-build to TypeScript

### LINKED ISSUE #2477: Migrate workbox-build to TypeScript
Placeholder to track the migration.

### LINKED ISSUE #2806: [workbox-build] Uses old deprecated @hapi/joi
Welcome! Please use this template for reporting bugs or requesting features. For questions about using Workbox, the best place to ask is Stack Overflow, tagged with `[workbox]`: https://stackoverflow.com/questions/ask?tags=workbox

**Library Affected**:
*workbox-build, workbox-webpack-plugin*

**Browser & Platform**:
*Webpack*

**Issue or Feature Request Description**:
*workbox-build is currently using a dep of @hapi/joi which is deprecated and just joi should be used now. Because it still uses @hapi/joi in version 16.* which is also old it has a dep to topo 3.x which is incompatible with the new 5.x version. Because of structuring of my packages it is now installing the wrong topo into my project. Would be cool to switch to the "real" dep.*

*When reporting bugs, please include relevant JavaScript Console logs and links to public URLs at which the issue could be reproduced.*

### PR DESCRIPTION
**Pull request by jeffposnick**
R: @tropicadri

Fixes #2806, fixes #2477

There's obviously a _lot_ going on in this PR. Sorry!

At a high level, this migrates `workbox-build` to TypeScript. (Following this PR, only `workbox-webpack-plugin` and `workbox-sw` are written in JavaScript.)

The goal was to preserve functionality during this migration, and not require a major semver bump. The biggest change under the hood is swapping out `@hapi/joi` option validation for a two-part system that generates JSON schema based on our TypeScript definitions at build time, and then uses https://ajv.js.org/ to validate at runtime against that schema. `@hapi/joi` is a bit more sophisticated in terms of complex runtime validation (e.g. ensuring that if one option is set, another option needs to also be set, making options mutually exclusive, etc.) so some of those relationships needed to be expressed at runtime via custom functions, meaning that the generated JSON schema are not 100% the source of truth.

Many of the test changes across all the packages reflect the fact that validation now throws slightly different errors due to the above mentioned change.

There are other small changes sprinkled throughout this PR as well, based on things that I found necessary while working with the additional TypeScript code. Please flag anything that looks amiss!

### CODE REVIEW

File: packages/workbox-build/src/lib/get-file-details.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
+/*
+  Copyright 2018 Google LLC
```
Comment (tropicadri) [inline packages/workbox-build/src/lib/get-file-details.ts]:
```suggestion
  Copyright 2021 Google LLC
```

File: test/workbox-build/node/generate-sw.js [hunk 2]
```diff
@@ -1066,33 +1034,23 @@ describe(`[workbox-build] generate-sw.js (End to End)`, function() {
...
-      } catch (error) {
-        expect(error.name).to.eql('ValidationError', error.message);
-        expect(error.details[0].context.key).to.eql('runtimeCaching');
-      }
+      await expect(generateSW(options)).to.eventually.be.rejectedWith(
+        WorkboxConfigError, errors['nav-preload-runtime-caching']);
     });
 
-    it(`should reject with a ValidationError when 'navigationPreload' is true and 'runtimeCaching' is invalid`, async function() {
+    it(`should reject when 'navigationPreload' is true and 'runtimeCaching' is invalid`, async function() {
```
Comment (tropicadri) [inline test/workbox-build/node/generate-sw.js]:
```suggestion
    it(`should reject when 'navigationPreload' is true and 'runtimeCaching' is undefined`, async function() {
```

File: packages/workbox-build/src/lib/additional-manifest-entries-transform.ts [hunk 3]
```diff
@@ -6,23 +6,30 @@
...
 
-const errors = require('./errors');
+import {errors} from './errors';
+import {ManifestEntry} from '../types';
 
-module.exports = (additionalManifestEntries) => {
-  return (manifest) => {
+export function additionalManifestEntriesTransform(additionalManifestEntries: Array<ManifestEntry | string>) {
+  return (manifest: Array<ManifestEntry & {size: number}>) => {
     const warnings = [];
```
Comment (tropicadri) [inline packages/workbox-build/src/lib/additional-manifest-entries-transform.ts]:
```suggestion
    const warnings : Array<string> = [];
```

File: packages/workbox-build/src/lib/additional-manifest-entries-transform.ts [hunk 4]
```diff
@@ -6,23 +6,30 @@
...
     const stringEntries = new Set();
```
Comment (tropicadri) [inline packages/workbox-build/src/lib/additional-manifest-entries-transform.ts]:
```suggestion
    const stringEntries = new Set<string>();
```

File: packages/workbox-build/src/lib/cdn-utils.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
+/*
+  Copyright 2018 Google LLC
```
Comment (tropicadri) [inline packages/workbox-build/src/lib/cdn-utils.ts]:
```suggestion
  Copyright 2021 Google LLC
```

File: packages/workbox-build/src/lib/get-file-manifest-entries.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
+/*
+  Copyright 2018 Google LLC
```
Comment (tropicadri) [inline packages/workbox-build/src/lib/get-file-manifest-entries.ts]:
```suggestion
  Copyright 2021 Google LLC
```

File: packages/workbox-build/src/lib/stringify-without-comments.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
+/*
+  Copyright 2018 Google LLC
```
Comment (tropicadri) [inline packages/workbox-build/src/lib/stringify-without-comments.ts]:
```suggestion
  Copyright 2021 Google LLC
```

File: packages/workbox-build/src/lib/validate-options.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
+/*
+  Copyright 2019 Google LLC
```
Comment (tropicadri) [inline packages/workbox-build/src/lib/validate-options.ts]:
```suggestion
  Copyright 2021 Google LLC
```

**Review by tropicadri — CHANGES_REQUESTED** [review]
There are several linting errors and I added a couple of suggestions.

**Comment by jeffposnick** [comment]
Thanks! Regarding the linting errors, I'm not going to be able to resolve the three instances of `Parsing error: Cannot read property 'map' of undefined` in this PR. As per https://stackoverflow.com/a/64626605/385997,  we're going to have to upgrade our TypeScript eslint dependencies to resolve them, and that's a larger effort being done in the other PR that you're working on.

The other errors should be resolved now, though.

File: packages/workbox-build/src/lib/additional-manifest-entries-transform.ts [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+  license that can be found in the LICENSE file or at
+  https://opensource.org/licenses/MIT.
+*/
+
+import {errors} from './errors';
+import {ManifestEntry} from '../types';
+
+export function additionalManifestEntriesTransform(additionalManifestEntries: Array<ManifestEntry | string>) {
+  return (manifest: Array<ManifestEntry & {size: number}>) => {
+    const warnings : Array<string> = [];
```
Comment (tropicadri) [inline packages/workbox-build/src/lib/additional-manifest-entries-transform.ts]:
```suggestion
    const warnings: Array<string> = [];
```

**Review by tropicadri — APPROVED** [review]
Sorry there was a typo in one of my previous suggestions, I added the fix.

Ack on the map undefined issue.

LGTM

**Comment by alsayedzomrawi** [comment]
[1394web.pdf](https://github.com/GoogleChrome/workbox/files/6833794/1394web.pdf)

**Pull request merged**